src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/DataAnnotationValidateOptions.cs

```diff
@@ -52,19 +55,86 @@ public ValidateOptionsResult Validate(string? name, TOptions options)
             ThrowHelper.ThrowIfNull(options);
 
             var validationResults = new List<ValidationResult>();
-            if (Validator.TryValidateObject(options, new ValidationContext(options), validationResults, validateAllProperties: true))
+            HashSet<object>? visited = null;
+            List<string>? errors = null;
+
+            if (TryValidateOptions(options, options.GetType().Name, validationResults, ref errors, ref visited))
             {
                 return ValidateOptionsResult.Success;
             }
 
-            string typeName = options.GetType().Name;
-            var errors = new List<string>();
-            foreach (ValidationResult result in validationResults)
+            Debug.Assert(errors is not null && errors.Count > 0);
+
+            return ValidateOptionsResult.Fail(errors);
+        }
+
+        [RequiresUnreferencedCode("This method on this type will walk through all properties of the passed in options object, and its type cannot be " +
+            "statically analyzed so its members may be trimmed.")]
+        private static bool TryValidateOptions(object options, string qualifiedName, List<ValidationResult> results, ref List<string>? errors, ref HashSet<object>? visited)
+        {
+            if (visited is not null && visited.Contains(options))
             {
-                errors.Add($"DataAnnotation validation failed for '{typeName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                return true;
             }
 
-            return ValidateOptionsResult.Fail(errors);
+            results?.Clear();
+
+            bool res = Validator.TryValidateObject(options, new ValidationContext(options), results, validateAllProperties: true);
+            if (!res)
+            {
+                errors ??= new List<string>();
+
+                foreach (ValidationResult result in results!)
+                {
+                    errors.Add($"DataAnnotation validation failed for '{qualifiedName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                }
+            }
+
+            foreach (PropertyInfo propertyInfo in options.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public))
+            {
+                if (propertyInfo.GetMethod is null)
+                {
+                    continue;
+                }
+
+                object? value = propertyInfo!.GetValue(options);
+
+                if (value is null)
+                {
+                    continue;
+                }
+
+                if (propertyInfo.GetCustomAttribute<ValidateObjectMembersAttribute>() is not null)
+                {
+                    visited ??= new HashSet<object>(new ReferenceComparer());
+                    visited.Add(options);
+
+                    results ??= new List<ValidationResult>();
```

Comment:
Just to match the same spacing used in the else if below
```suggestion
                    results ??= new List<ValidationResult>();

```